s4/clarity/scripts/stepepp.py

```diff
@@ -134,7 +134,7 @@ def prefetch(self, *categories):
         if StepEPP.PREFETCH_SAMPLES in categories:
             # use the artifacts that we prefetched to find samples.
             # we use a set because we may have both inputs and outputs in the artifacts_to_fetch list.
-            samples = list(set([x.sample for x in artifacts]))
+            samples = list(set([sample for a in artifacts for sample in a.samples]))
```

Comment:
I only changed `x` to `a` here because I thought that using `a` for the artifact might be a helpful hint in the longer expression. I can leave it as `x` if that doesn't make sense to anyone else. :)

Reply:
Oh, interesting. Is `sample for a in artifacts for sample in a.samples` a generator expression in that case?

Any preference between your suggestion `list(set(sample for a in artifacts for sample in a.samples))` and a set comprehension `list({sample for a in artifacts for sample in a.samples})`? I discovered set comprehensions while trying to find out what a "list comprehension without []" is called. :)